lib/logger/lib/logger.ex

```diff
@@ -671,7 +679,7 @@ defmodule Logger do
   @doc false
   def __do_log__({level, config, pdict}, chardata_or_fun, metadata) when is_list(metadata) do
     %{utc_log: utc_log?, truncate: truncate, mode: mode} = config
-    metadata = [pid: self()] ++ Keyword.merge(pdict, metadata)
+    metadata = [pid: self()] ++ into_metadata(metadata, pdict)
```

Comment:
Micro-micro optimization: `[{:pid, self()} | into_metadata(metadata, pdict)]`?

Reply:
They are equivalent when the left side of `++` is a literal.

Reply:
Interesting. Is this documented? If not we should

Reply:
It is a compiler optimisation. There are many of them and they also change between versions - I don't think it would be useful to document them all.

In general, one should write the most straightforward code and expect the compiler to translate it to the most efficient code.

Reply:
Sorry, my Erlarng-literals-fu is weak 😅 

I was wondering how `[pid: self()]` can be treated as a literal if `self()` is obtained at runtime. Is there any compiler optimization magic ✨ I can look into?

(Sorry for using this for learning purposes instead of the forum)

Reply:
It's a "partial" literal - since some structure is known (the length of the list in particular), the compiler can emit some better code than first building the entire list and then calling the append operation.